File: split/beginbox.pas
procedure beginbox(boxcontext: integer);
label
  10, 30;
var
  p, q: halfword;
  m: quarterword;
  k: halfword;
  n: eightbits;
begin
  case curchr of
    0:
    begin
      scaneightbitint;
      curbox := eqtb[3678 + curval].hh.rh;
      eqtb[3678 + curval].hh.rh := 0;
    end;
    1:
    begin
      scaneightbitint;
      curbox := copynodelist(eqtb[3678 + curval].hh.rh);
    end;
    2:{1080:}

    begin
      curbox := 0;
      if abs(curlist.modefield) = 203 then
      begin
        youcant;
        begin
          helpptr := 1;
          helpline[0] := 1069;
        end;
        error;
      end
      else if (curlist.modefield = 1) and (curlist.headfield = curlist.tailfield) then
      begin
        youcant;
        begin
          helpptr := 2;
          helpline[1] := 1070;
          helpline[0] := 1071;
        end;
        error;
      end
      else if not (curlist.tailfield >= himemmin) then
        if (mem[curlist.tailfield].hh.b0 = 0) or (mem[curlist.tailfield].hh.b0 = 1) then{1081:}

        begin
          q := curlist.headfield;
          repeat
            p := q;
            if not (q >= himemmin) then
              if mem[q].hh.b0 = 7 then
              begin
                for m := 1 to mem[q].hh.b1 do
                  p := mem[p].hh.rh;
                if p = curlist.tailfield then
                  goto 30;
              end;
            q := mem[p].hh.rh;
          until q = curlist.tailfield;
          curbox := curlist.tailfield;
          mem[curbox + 4].int := 0;
          curlist.tailfield := p;
          mem[p].hh.rh := 0;
          30: ;
        end{:1081}
;
    end{:1080}
;
    3:{1082:}

    begin
      scaneightbitint;
      n := curval;
      if not scankeyword(841) then
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(1072);
        end;
        begin
          helpptr := 2;
          helpline[1] := 1073;
          helpline[0] := 1074;
        end;
        error;
      end;
      scandimen(false, false, false);
      curbox := vsplit(n, curval);
    end{:1082}
;
    others:{1083:}

    begin
      k := curchr - 4;
      savestack[saveptr + 0].int := boxcontext;
      if k = 102 then
        if (boxcontext < 1073741824) and (abs(curlist.modefield) = 1) then
          scanspec(3, true)
        else
          scanspec(2, true)
      else
      begin
        if k = 1 then
          scanspec(4,
            true)
        else
        begin
          scanspec(5, true);
          k := 1;
        end;
        normalparagraph;
      end;
      pushnest;
      curlist.modefield := -k;
      if k = 1 then
      begin
        curlist.auxfield.int := -65536000;
        if eqtb[3418].hh.rh <> 0 then
          begintokenlist(eqtb[3418].hh.rh, 11);
      end
      else
      begin
        curlist.auxfield.hh.lh := 1000;
        if eqtb[3417].hh.rh <> 0 then
          begintokenlist(eqtb[3417].hh.rh, 10);
      end;
      goto 10;
    end{:1083}

  end;
  boxend(boxcontext);
  10: ;
end;

File: cpp/beginbox.cpp
#include "beginbox.h"
#include "scaneightbitint.h"
#include "copynodelist.h"
#include "youcant.h"
#include "error.h"
#include "scankeyword.h"
#include "printnl.h"
#include "print.h"
#include "vsplit.h"
#include "scandimen.h"
#include "normalparagraph.h"
#include "scanspec.h"
#include "pushnest.h"
#include "begintokenlist.h"
#include "boxend.h"
#include "texte.h"

void beginbox(int boxcontext)
{
	halfword p, q;
	quarterword m;
	halfword k;
	eightbits n;
	switch (curchr)
	{
		case 0:
			scaneightbitint();
			curbox = box(curval);
			box(curval) = 0;
			break;
		case 1:
			scaneightbitint();
			curbox = copynodelist(box(curval));
			break;
		case 2:
			curbox = 0;
			if (abs(mode) == mmode)
			{
				youcant();
				helpptr = 1;
				helpline[0] = "Sorry; this \\lastbox will be void.";
				error();
			}
			else 
				if (mode == vmode && head == tail)
				{
					youcant();
					helpptr = 2;
					helpline[1] = "Sorry...I usually can't take things from the current page.";
					helpline[0] = "This \\lastbox will therefore be void.";
					error();
				}
				else 
					if (tail < himemmin)
						if (type(tail) == hlist_node || type(tail) == vlist_node)
						{
							q = head;
							bool brk = false;
							do
							{
								p = q;
								if (q < himemmin && type(q) == disc_node) // 7
								{
									for (m = 1 ; m <= subtype(q); m++)
										p = link(p);
									if (p == tail)
									{
										brk = true;
										break;
									}
								}
								q = link(p);
							} while (q != tail);
							if (!brk)
							{
								curbox = tail;
								shift_amount(curbox) = 0;
								tail = p;
								link(p) = 0;
							}
						}
			break;
		case 3:
			scaneightbitint();
			n = curval;
			if (!scankeyword("to"))
			{
				printnl("! ");
				print("Missing `to' inserted");
				helpptr = 2;
				helpline[1] = "I'm working on `\\vsplit<box number> to <dimen>';";
				helpline[0] = "will look for the <dimen> next.";
				error();
			}
			scandimen(false, false, false);
			curbox = vsplit(n, curval);
			break;
		default:
			k = curchr-4;
			savestack[saveptr].int_ = boxcontext;
			if (k == hmode)
				if (boxcontext < 0x40'00'00'00 && abs(mode) == vmode)
					scanspec(3, true);
				else
					scanspec(2, true);
			else
			{
				if (k == vmode)
					scanspec(4,	true);
				else
				{
					scanspec(5, true);
					k = vmode;
				}
				normalparagraph();
			}
			pushnest();
			mode = -k;
			if (k == vmode)
			{
				prev_depth = ignore_depth;
				if (every_vbox())
					begintokenlist(every_vbox(), every_vbox_text);
			}
			else
			{
				space_factor = 1000;
				if (every_hbox())
					begintokenlist(every_hbox(), every_hbox_text);
			}
			return;
	}
	boxend(boxcontext);
}
